### DISCUSSION TITLE
How to tell linq2db that all table names are lowercase

### DISCUSSION THREAD

**Discussion by jonreis** [linq2db/linq2db#4634] [Q&A] [answered]
I am new to linq2db and I am attempting to use existing POCO object between 2 databases: Sql Server and Postgres. The POCO objects are Pascal cased. I have used the scaffolding tool to generate a Fluent data model from my Sql Server database.

Aside: It would have been very helpful to have a 

public virtual void BeforeSourceCodeGenerated(FinalDataModel model) where I could adjust all aspects of the generated model including the generated source code.

I would like to use the same POCO model against my Postgres database but I am having trouble with the casing as all table and column names are lower case in the Postgres database. I found this:

    var options = new DataOptions();
    options = options.UsePostgreSQL(DatabaseConnectionString).WithOptions(new PostgreSQLOptions(BulkCopyType.MultipleRows, true, PostgreSQLIdentifierQuoteMode.None));

But, I would prefer to just tell linq2db that all the table names are lowercase so that I can use the default PostgreSQLIdentifierQuoteMode = Auto

I have tried registering a IMetadataReader

    LinqToDB.Mapping.MappingSchema.Default.AddMetadataReader(new CaseInsensitiveMetadataReader());

But this seems to trigger reflection of the POCO objects and reads in properties that are not part of the database model.

Can someone please give me some advice on how I can tell linq2db that all table and column names for Posgres should be assumed to be lowercase?

```
public class CaseInsensitiveMetadataReader : IMetadataReader
{
  private readonly AttributeReader _reader = new();

  public MappingAttribute[] GetAttributes(Type type)
  {
    var attributes = _reader.GetAttributes(type);

    if (attributes.Any(x => x is TableAttribute))
    {
      return attributes;
    }

    return attributes.Concat([
      new TableAttribute()
      {
        // get type name and apply naming convention
        Name = type.Name.ToLowerInvariant()
      }
    ]).ToArray();
  }

  public MappingAttribute[] GetAttributes(Type type, MemberInfo memberInfo)
  {
    var attributes = _reader.GetAttributes(type);

    if (attributes.Any(x => x is ColumnAttribute))
    {
      return attributes;
    }

    return attributes.Concat([
      new ColumnAttribute()
      {
        // get type name and apply naming convention
        Name = memberInfo.Name.ToLowerInvariant(),
      }
    ]).ToArray();
  }

  public MemberInfo[] GetDynamicColumns(Type type) => [];

  public string GetObjectID()
  {
    return nameof(CaseInsensitiveMetadataReader);
  }
}
```

**Comment by MaceWindu** [comment, marked as answer]
Custom `IMetadataReader`  is a recommended way, you just need to replace `AttributeReader` instance with `FluentMetadataReader` instance, created by fluent mapping builder. But I see that there is no easy way to get this instance.

As alternative I would propose to use entity descriptor creation callback and edit entity descriptor, generated from mappings. 

```cs
MappingSchema.EntityDescriptorCreatedCallback = (ms, ed) =>
{
    if (ms == expectedMappingSchema)
    {
        ed.TableName = ed.TableName.ToLowerInvariant();
        foreach (var col in ed.Columns)
        {
            col.ColumnName = col.ColumnName.ToLowerInvariant();
        }
    }
}
```

**Reply by jonreis** [reply]
@MaceWindu that did the trick. Thanks for your help with this! I was banging my head trying to figure it out.
